pyrt/renderer/simplert.py

```diff
@@ -69,22 +69,39 @@ def _shadow(self, scene: Scene, hitrecord: HitRecord) -> tuple:
             fs /= 4.
 
         return fs,local_num_shadow_rays
+    
+    def _recurse_shade(self, scene: Scene, ray: Ray, iteration_num: int) -> tuple:
+        hitrecord = HitRecord()
+        hit, r, g, b = self._shade(scene, ray, hitrecord)
 
-    def _reflect(self, r: int, g: int, b: int, scene: Scene, ray: Ray, hitrecord: HitRecord) -> tuple:
-        reflect_ray = Ray(hitrecord.point, reflect3(hitrecord.normal_g, ray.direction))
-        reflect_hitrecord = HitRecord()
-
-        hit, rnew, gnew, bnew = self._shade(scene, reflect_ray, reflect_hitrecord)
+        if not hit or iteration_num == 0:
+            return r, g, b
+        
+        if hitrecord.material.reflectivity != 0.0:
+            reflect_ray = Ray(hitrecord.point, reflect3(hitrecord.normal_g, ray.direction))
+            new_r, new_g, new_b = self._recurse_shade(scene, reflect_ray, iteration_num - 1)
 
-        if hit:
             ref1 = hitrecord.material.reflectivity
             ref2 = 1.-ref1
-            rnew = int ( ref1 * rnew + ref2 * r)
-            gnew = int ( ref1 * gnew + ref2 * g)
-            bnew = int ( ref1 * bnew + ref2 * b)
-            return rnew, gnew, bnew, reflect_ray, reflect_hitrecord
-        else:
-            return r,g,b,None,None
+            r = int ( ref1 * new_r + ref2 * r)
+            g = int ( ref1 * new_g + ref2 * g)
+            b = int ( ref1 * new_b + ref2 * b)
+
+            self.num_secondary_rays += 1
+
+        if hitrecord.material.transparency != 0.0:
```

Comment:
Should be a fuzzy EPSILON comparison

Be careful when using floating-point numbers

-EPS < a < EPS